File: android/app/src/main/kotlin/org/gaindrive/android/data/model/BrowseScope.kt
package org.gaindrive.android.data.model

/**
 * Which servers the browse screens are showing.
 *
 * [AllServers] is the default rather than a named server: with one server
 * configured the two are the same thing and the selector is hidden entirely, so
 * the app still looks like a single-server client until it isn't one. With
 * several, seeing all of them is the reason they were added.
 */
sealed interface BrowseScope {

	data object AllServers : BrowseScope

	data class OneServer(val id: ServerId) : BrowseScope

	companion object {
		/** Stored in place of a server id. Not a UUID, so it cannot collide. */
		const val ALL_STORED = "all"
	}
}

/**
 * Everything that decides what a browse screen shows.
 *
 * The two travel together because a screen has to reload for either: picking a
 * different server is a different library, and going offline is the same
 * library from a different source.
 */
data class BrowseState(
	val scope: BrowseScope,
	val offline: Boolean,
	/**
	 * Bumped on every change to the server list — see `ServerRegistry.revision`,
	 * which explains why the scope alone cannot carry one.
	 *
	 * Mostly it is here so that a change compares unequal and the screens
	 * reload. The number itself is only read where a screen distinguishes
	 * "reload" from "reload and blank what is there first", as the playlists
	 * screen does.
	 */
	val revision: Int = 0,
)

File: ios/GainDrive/Data/Model/BrowseScope.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// Which servers a browse screen is showing.
///
/// `allServers` is the default rather than a named server: with one server
/// configured the two are the same thing and the selector is hidden, so **the
/// app still looks like a single-server client until it isn't one**; with
/// several, seeing all of them is the reason they were added.
enum BrowseScope: Hashable, Sendable {
	case allServers
	case oneServer(ServerId)

	/// The stored form. A sentinel rather than an empty string or a flag,
	/// because it has to share a field with a `ServerId` - and `"all"` is not
	/// a UUID, so the two can never collide.
	static let allStored = "all"

	var stored: String {
		switch self {
		case .allServers: Self.allStored
		case .oneServer(let id): id.value.uuidString
		}
	}

	/// Resolves a stored choice against the servers that currently exist.
	///
	/// A choice whose server was removed or disabled falls back to all rather
	/// than leaving the library permanently empty with no hint why - which is
	/// what a scope pointing at a server that is not there would do.
	static func restored(from stored: String?, available: [ServerId]) -> BrowseScope {
		guard let stored, stored != allStored,
			let uuid = UUID(uuidString: stored),
			available.contains(ServerId(uuid))
		else {
			return .allServers
		}
		return .oneServer(ServerId(uuid))
	}
}
